tar_sam:
  #--->Topics
  pointCloudTopic: "velodyne_points"
  imuTopic: "/xsens/imu/data"
  gpsTopic: "odometry/gpsz"                   # GPS odometry topic from navsat, see module_navsat.launch file

  #--->Frames
  lidarFrame: velodyne
  odomFrame: odom
  mapFrame: map
  baselinkFrame: base_link


  #--->Sensor Settings
  sensor: velodyne                            #Lidar sensor type: velodyne 
  N_SCAN: 16                                  #Number of lidar scans: 16 
  SCAN_Period: 0.1                            #Scan period 
  Horizon_SCAN: 1800                          #Lidar horizontal resolution: 1800
  lidarMinRange: 1.0                          #Minimum lidar range to be used
  lidarMaxRange: 1000.0                       #Maximum lidar range to be used
   
  #---> IMU Settings
  imuAccNoise: 3.9939570888238808e-03
  imuGyrNoise: 1.5636343949698187e-03
  imuAccBiasN: 6.4356659353532566e-05
  imuGyrBiasN: 3.5640318696367613e-05
  imuGravity: 9.80511
  imuRPYWeight: 0.01

  # Extrinsics: T_lb (lidar -> imu)
  extrinsicTrans: [0.0, 0.0, 0.0]
  extrinsicRot: [1, 0, 0,
                  0, -1, 0,
                  0, 0, -1]
  extrinsicRPY: [1, 0, 0,
                 0, -1, 0,
                 0, 0, -1]
  # extrinsicRot: [1, 0, 0,
  #                 0, 1, 0,
  #                 0, 0, 1]
  # extrinsicRPY: [1, 0, 0,
  #                 0, 1, 0,
  #                 0, 0, 1]

  #--->Fast Odometry Graph
    #LiDAR Odometry
  DISTANCE_SQ_THRESHOLD: 25
  NEARBY_SCANEARBY_SCAN: 2.5
  skipFrameNum: 5

  #---> Map Optimization

  # GPS Settings
  useImuHeadingInitialization: true           # if using GPS data, set to "true"
  useGpsElevation: true                      # if GPS elevation is bad, set to "false"
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 25.0                      # m^2, threshold for using GPS data

  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # voxel filter paprams
  mappingCornerLeafSize: 0.2                    # default: 0.2 - outdoor, 0.1 - indoor
  mappingSurfLeafSize: 0.4                      # default: 0.4 - outdoor, 0.2 - indoor

  # robot motion constraint (in case you are using a 2D robot)
  z_tollerance: 1000                            # meters
  rotation_tollerance: 1000                     # radians

  # CPU Params
  numberOfCores: 4                              # number of cores for mapping optimization
  mappingProcessInterval: 0.15                  # seconds, regulate mapping frequency

  # Surrounding map
  surroundingkeyframeAddingDistThreshold: 0.5   # meters, regulate keyframe adding threshold
  surroundingkeyframeAddingAngleThreshold: 0.01  # radians, regulate keyframe adding threshold
  surroundingKeyframeDensity: 1.0               # meters, downsample surrounding keyframe poses   
  surroundingKeyframeSearchRadius: 100.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)

  # Loop closure
  loopClosureEnableFlag: true
  loopClosureFrequency: 1.0                     # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 15.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0           # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
  historyKeyframeFitnessScore: 0.3              # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density






